File: homework4/phuc_le/q1_BuildTrie.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>

using namespace std;

struct TrieNode
{
    vector<TrieNode *> children;
    bool validWord;

    TrieNode()
    {
        children = vector<TrieNode *>(26, nullptr);
        validWord = false;
    }
};

class Trie
{
private:
    TrieNode *root;

public:
    Trie()
    {
        root = new TrieNode();
    }

    /*
        Inserts a word into the trie.
        - Start at the root node.
        - For each character, compute its index (0-25).
        - If the child node does not exist, create it.
        - Move to the child node.
        - After processing all characters, mark the last node as a valid word.

        Time: O(n), where n is the length of the word.
        Space: O(n) worst case if new nodes are created.
    */
    void insert(const string &word)
    {
        TrieNode *curr = root;

        for (char c : word)
        {
            int idx = c - 'a';

            if (curr->children[idx] == nullptr)
                curr->children[idx] = new TrieNode();

            curr = curr->children[idx];
        }

        curr->validWord = true;
    }

    /*
    Returns True if the word exists in the trie.
        - Start at the root node.
        - For each character, compute its index.
        - If the child does not exist, return False.
        - Move to the child node.
        - After processing all characters, return whether the final node
        is marked as a valid word.

        Time: O(n), where n is the length of the word.
        Space: O(1).
    */
    bool isValidWord(const string &word)
    {
        TrieNode *curr = root;

        for (char c : word)
        {
            int idx = c - 'a';

            if (curr->children[idx] == nullptr)
                return false;

            curr = curr->children[idx];
        }

        return curr->validWord;
    }

    /*
        Recursively removes a word from the trie.

        Returns:
        true  -> this node is no longer needed and can be deleted.
        false -> this node must remain in the trie.

        Time: O(n)
        Space: O(n)
    */
    bool removeHelper(TrieNode *node, const string &word, int depth)
    {
        if (node == nullptr)
            return false;

        // Reached end of word
        if (depth == word.size())
        {
            if (!node->validWord)
                return false;
            node->validWord = false;
            // If node has no children, delete it
            for (auto child : node->children)
            {
                if (child != nullptr)
                    return false;
            }
            return true;
        }

        int idx = word[depth] - 'a';
        TrieNode *child = node->children[idx];

        if (child == nullptr)
            return false;

        bool shouldDeleteChild = removeHelper(child, word, depth + 1);

        if (shouldDeleteChild)
        {
            delete child;
            node->children[idx] = nullptr;
        }

        // Never delete root
        if (node == root)
            return false;
        if (node->validWord)
            return false;

        for (auto c : node->children)
        {
            if (c != nullptr)
                return false;
        }

        return true;
    }

    /*
        Removes a word from the trie.
        - Use recursion to reach the end of the word.
        - If the word does not exist, do nothing.
        - Unmark the last node as a valid word.
        - While backtracking, delete nodes that:
            1. Have no children.
            2. Are not the end of another word.
        - Stop deleting once a node is still needed.

        Time: O(n), where n is the length of the word.
        Space: O(n) due to the recursion stack.
    */
    void remove(const string &word)
    {
        removeHelper(root, word, 0);
    }
};

// Tests

int main()
{
    Trie t;

    // Empty trie
    assert(!t.isValidWord("apple"));

    // Insert one word
    t.insert("apple");
    assert(t.isValidWord("apple"));
    assert(!t.isValidWord("app"));

    // Insert multiple words
    t.insert("app");
    t.insert("banana");
    assert(t.isValidWord("app"));
    assert(t.isValidWord("banana"));

    // Remove a word
    t.remove("app");
    assert(!t.isValidWord("app"));
    assert(t.isValidWord("apple"));

    // Remove another word
    t.remove("apple");
    assert(!t.isValidWord("apple"));
    assert(t.isValidWord("banana"));

    // Remove non-existing word
    t.remove("cat");
    assert(t.isValidWord("banana"));

    // Single-character word
    t.insert("a");
    assert(t.isValidWord("a"));
    t.remove("a");
    assert(!t.isValidWord("a"));

    cout << "All tests passed!\n";
    return 0;
}

File: homework4/aman_varshney/q1_BuildTrie.py
# spent 45 minutes

class TrieNode:
    def __init__(self) -> None:
        self.children = {} # size of 26
        self.valid_word = False # indicates if this node marks the end of a word
    
    
class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()
        
        
    def insert(self, word: str) -> None:
        """Adds a word to the trie"""
        # Add word
        curr = self.root
        for char in word: 
            if char not in curr.children: # Create path if DNE
                curr.children[char] = TrieNode()
                
            curr = curr.children[char] # Follow path in trie
            
        # Mark as valid word
        curr.valid_word = True
        
        
    def is_valid_word(self, word: str) -> bool: 
        """Returns true if word is ADDED in trie; otherwise false"""
        # Iterate through trie using word characters
        curr = self.root
        for char in word:
            if char not in curr.children: # Fail if path not found
                return False
            curr = curr.children[char]
            
        return curr.valid_word
        
        
    def remove(self, word: str) -> None:
        """Removes a word from the trie and deletes the unused nodes"""
        
        # helper
        def dfs(node: TrieNode, idx: int) -> bool:
            """Returns True if `node` should be deleted (unused branch)."""
            # Reached end of word
            if idx == len(word):
                # Unmark the word if it exists
                if not node.valid_word:
                    return False
                node.valid_word = False
                # Mark for deletion only if no words continue further
                return len(node.children) == 0 
            
            char = word[idx]
            if char not in node.children: # Path DNE
                return False
            
            
            # Delete edge if child branch is now unused
            child = node.children[char]
            if dfs(child, idx+1):
                del node.children[char]
            
            # Delete parent if it has no children and is not a word
            return len(node.children) == 0 and not node.valid_word
        
    
        dfs(self.root, 0)



def run_tests() -> None:
    print("Running tests")
    trie = Trie()

    # Empty tests: is_valid_word()
    assert not trie.is_valid_word("cat")
    trie.remove("cat")


    # insert() and valid_word tests
    trie.insert("cat")
    assert trie.is_valid_word("cat")

    trie.insert("car")
    trie.insert("card")
    trie.insert("catch")

    assert trie.is_valid_word("cat")
    assert trie.is_valid_word("car")
    assert trie.is_valid_word("card")
    assert trie.is_valid_word("catch")
    assert not trie.is_valid_word("ca")
    assert not trie.is_valid_word("random")
    assert not trie.is_valid_word("cardx")

    
    # remove() tests
    # Prefix words
    trie.remove("car")
    assert not trie.is_valid_word("car")
    assert trie.is_valid_word("card")
    assert trie.is_valid_word("cat")
    assert trie.is_valid_word("catch")

    # Nonexistent word
    trie.remove("dne")
    assert trie.is_valid_word("cat")
    assert trie.is_valid_word("card")
    assert trie.is_valid_word("catch")
    assert not trie.is_valid_word("dne")

    # Suffix words
    trie.insert("car")
    trie.remove("card")
    assert not trie.is_valid_word("card")
    assert trie.is_valid_word("car")

    # Full path cleanup
    trie.remove("cat")
    trie.remove("catch")
    trie.remove("car")
    assert not trie.is_valid_word("cat")
    assert not trie.is_valid_word("car")
    assert not trie.is_valid_word("card")
    assert not trie.is_valid_word("catch")
    assert not trie.is_valid_word("ca")
    assert not trie.is_valid_word("c")


    print()
    print("PASS")


if __name__ == "__main__":
    run_tests()
